src/pipx/shared_libs.py

```diff
@@ -32,7 +32,12 @@ def create(self, pip_args: List[str], verbose: bool = False):
         if not self.is_valid:
             with animate("creating shared libraries", not verbose):
                 run([DEFAULT_PYTHON, "-m", "venv", "--clear", self.root])
-            self.upgrade(pip_args, verbose)
+            # Note: -I is needed here to replace the patched Debian pip with
+            # the upstream one, even when the Debian pip version matches the
+            # current upstream pip version.  Otherwise there's a flood of
+            # bug reports about things not working, like
+            # https://github.com/pipxproject/pipx/issues/386
+            self.upgrade(["-I"] + pip_args, verbose)
```

Comment:
Will this be a NO OP when running outside of Debian? Would not want to make all platforms slower to patch a Debian bug.

Reply:
I don't know.  Some measurements would be nice to have.  I don't have any non-Debian-based systems to test against, unless maybe I could come up with something using Docker and Alpine?

Reply:
lol, -I makes this step _faster_

```
$ docker run -ti --rm python:3.8-alpine sh
... some experiments to get the commands right ...
... which incidentally populate ~/.pip/cache so the measurements are independent of network speed ...
/ # rm -rf env && time python3 -m venv env --clear && time env/bin/python -m pip --disable-pip-version-check install --upgrade pip setuptools wheel
real	0m 1.75s
user	0m 1.59s
sys	0m 0.15s
Collecting pip
  Using cached https://files.pythonhosted.org/packages/54/0c/d01aa759fdc501a58f431eb594a17495f15b88da142ce14b5845662c13f3/pip-20.0.2-py2.py3-none-any.whl
Collecting setuptools
  Using cached https://files.pythonhosted.org/packages/a0/df/635cdb901ee4a8a42ec68e480c49f85f4c59e8816effbf57d9e6ee8b3588/setuptools-46.1.3-py3-none-any.whl
Collecting wheel
  Using cached https://files.pythonhosted.org/packages/8c/23/848298cccf8e40f5bbb59009b32848a4c38f4e7f3364297ab3c3e2e2cd14/wheel-0.34.2-py2.py3-none-any.whl
Installing collected packages: pip, setuptools, wheel
  Found existing installation: pip 19.2.3
    Uninstalling pip-19.2.3:
      Successfully uninstalled pip-19.2.3
  Found existing installation: setuptools 41.2.0
    Uninstalling setuptools-41.2.0:
      Successfully uninstalled setuptools-41.2.0
Successfully installed pip-20.0.2 setuptools-46.1.3 wheel-0.34.2
real	0m 2.53s
user	0m 2.05s
sys	0m 0.19s
/ # rm -rf env && time python3 -m venv env --clear && time env/bin/python -m pip --disable-pip-version-check install -I --upgrade pip setuptools wheel
real	0m 1.77s
user	0m 1.61s
sys	0m 0.15s
Collecting pip
  Using cached https://files.pythonhosted.org/packages/54/0c/d01aa759fdc501a58f431eb594a17495f15b88da142ce14b5845662c13f3/pip-20.0.2-py2.py3-none-any.whl
Collecting setuptools
  Using cached https://files.pythonhosted.org/packages/a0/df/635cdb901ee4a8a42ec68e480c49f85f4c59e8816effbf57d9e6ee8b3588/setuptools-46.1.3-py3-none-any.whl
Collecting wheel
  Using cached https://files.pythonhosted.org/packages/8c/23/848298cccf8e40f5bbb59009b32848a4c38f4e7f3364297ab3c3e2e2cd14/wheel-0.34.2-py2.py3-none-any.whl
Installing collected packages: pip, setuptools, wheel
Successfully installed pip-20.0.2 setuptools-46.1.3 wheel-0.34.2
real	0m 1.83s
user	0m 1.47s
sys	0m 0.18s
```

without -I: 2.53s
with -I: 1.83s

overhead of python -m venv env --clear: 1.75s (same in both cases)

Reply:
> which incidentally populate ~/.pip/cache so the measurements are independent of network speed ...

Unless I'm misunderstanding something - this comment highlights that time measured avoids one of the most relevant impacts - all users will have to download pip. Here are the docs for `-I`

```
-I, --ignore-installed      Ignore the installed packages, overwriting them. This can 
                              break your system if the existing package is of a different version
                              or was installed with a different package manager!
```
So this would *always* ignore the system pip (possibly re-downloading it) for everyone.

Reply:
That is a good point, but as you saw in my terminal log, pip is downloaded and installed irrespective of the -I flag.

I'm not familiar with venv and I don't know where it found the old outdated versions of pip/setuptools that got installed in my experimental venv.  Do you know how I could reproduce (in Docker?) a situation where venv installs up-to-date versions if pip/setuotools that don't require an upgrade?

Reply:
> pip is downloaded and installed irrespective of the -I flag.
> That is a good point, but as you saw in my terminal log, pip is downloaded and installed irrespective of the -I flag.

The log shows pip was cached, but in any case, the shared libs directory is created rarely, and installing a version of pip that is guaranteed to be good is appealing. Being able to install packages from the internet is required for pipx to work, so I don't think pip, setuptools, and wheel being downloaded every once in a while will have a significant negative impact on user experience.

Is this actually overwriting the system pip or just the pip in the venv? If it's the former I don't think we should land it because we could be breaking OS's. If it's the latter, I think we should merge this since it will make pipx hermetic to whatever weird system things are going on. Does anyone know?

Reply:
Another q: should `--force-reinstall` be preferred? https://stackoverflow.com/questions/51913361/difference-between-pip-install-options-ignore-installed-and-force-reinstall indicated the follwing

> "Otherwise, you may get obscure import errors after reinstall due to stale modules still available in sys.path."

Reply:
There is a scary warning in the pip docs:

> Ignore the installed packages, overwriting them. This can break your system if the existing package is of a different version or was installed with a different package manager!

https://pip.pypa.io/en/stable/reference/pip_install/

I don't know enough about the relationship between OS packages vs how `--ignore-installed` or `--force-reinstall` affects them, so unfortunately I don't feel comfortable merging this.